src/data/tprod.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2020 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import data.list.nodup
+
+/-!
+# Finite products of types
+
+This file defines the product of types over a list. For `l : list ι` and `α : ι → Type*` we define
+`list.tprod α l = l.foldr (λ i β, α i × β) punit`.
+This type should not be used if `Π i, α i` or `Π i ∈ l, α i` can be used instead
+(in the last expression, we could also replace the list `l` by a set or a finset).
+This type is used as an intermediary between binary products and finitary products.
+The application of this type is finitary product measures, but it could be used in any
+construction/theorem that is easier to define/proof on binary products than on finitary products.
+
+* Once we have the construction on binary products (like binary product measures in
+  `measure_theory.prod`), we can easily define a finitary version on the type `tprod l α`
+  by iterating. Properties can also be easily extended from the binary case to the finitary case
+  by iterating.
+* Then we can use the equivalence `list.tprod.pi_equiv_tprod` below (or enhanced versions of it,
+  like a `measurable_equiv` for product measures) to get the construction on `Π i : ι, α i`, at
+  least when assuming `[fintype ι] [encodable ι]`.
+  Using choice or in proofs, we can remove the argument `[encodable I]` using
+  `encodable.trunc_encodable_of_fintype`.
```

Comment:
I just approved #5389, so maybe you can rephrase this using the more user-friendly `local attribute` version.